proposals/type-imports/Overview.md

```diff
@@ -52,67 +96,165 @@ For the following code,
+  - `heaptype ::= ... | any`
+  - the type of all importable (and referenceable) types
+
+
 #### Imports
 
-* `type <typetype>` is an import description with an upper bound
+* `type <typetype>` is an import description with a type constraint
   - `importdesc ::= ... | type <typetype>`
   - Note: `type` may get additional parameters in the future
 
-* `typetype` describes the type of a type import, as an upper bound
+* `sub <heaptype>` describes the type of a type import, with an upper bound
   - `typetype ::= sub <heaptype>`
   - `(sub <heaptype>) ok` iff `<heaptype> ok`
-  - Note: the bound can be a function type
   - Note: there may be other kinds of type descriptions in the future
-  - Note: there has to be an additional module-level side condition ensuring that bounds do not form a non-productive cycle
 
-* Type imports have indices prepended to the type index space, similar to other imports.
-  - Note: due to bounds, type imports can be mutually recursive with other type imports as well as regular type definitions. Hence they have to be validated together with the type section.
+* Type imports share the usual type index space, and are inserted in order of appearance.
+
+* There are additional side conditions on the ordering of type imports and definitions: they can be interleaved, but they are ordered; a type import may only be referenced by later declarations; this ensures that import bounds cannot form a non-productive definitional cycle; consecutive type definitions OTOH may be mutually recursive (the exact details of these rules are TBD and may be relaxed in future versions of Wasm; cf. Module Linking proposal).
+
+
+#### Definitions
+
+* `deftype` is a new category of *defined types* that generalises the contents of the type section
+  - `deftype ::= <functype> | <privatetype>`
+  - `module ::= {..., types vec(<deftype>)}`
+
+* `private <valtype>*` is a new form of type definition
+  - `privatetype ::= private <valtype>*`
+  - `private <valtype>* ok` iff `(<valtype> ok)*`
+  - private types are *nominal*, i.e., two structurally equivalent definitions produce distinct types
+  - This is similar to a nominal immutable struct, and could later be unified with the struct mechanism under the GC proposal (see below).
 
-Note: `<heaptype>` is defined in the [typed function references proposal](https://github.com/WebAssembly/function-references). It is either a type index or an abstract type like `func`, `extern`, etc.
 
 #### Exports
 
 * `type <heaptype>` is an export description
   - `exportdesc ::= ... | type <heaptype>`
   - `(type <heaptype>) ok` iff `<heaptype> ok`
+  - the definition of a every export type is transparently visible outside the module, except if it is defined as a private type
 
+Question: This does not allow exposing the definition of a private type to cooperating "friend" modules.
+As a more flexible alternative, hiding the definition could be optional via an explicit annotation on type exports.
```

Comment:
This does seem like a desirable ability; how do you see it working?  (Note also it'd be useful to be able to share a private type with a nested instance by passing it to `instantiate`.)  Running with the OOP-y naming (you started it ;), would it be some kind of `(type $T (protected $P))` definition, where `$P` is private, and then you can use `$T` when you do really mean to share?

Reply:
Ah, ok.